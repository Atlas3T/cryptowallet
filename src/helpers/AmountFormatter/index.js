/**
 * @TODO James Konrad
 * I've mocked some exchange rates below for the amount() function to use
 * Would make sense to put them somewhere in the Vuex store
 */

import numeral from 'numeral';

const exchangeRates = {
  BTC: [
    {
      currency: 'USD',
      rate: 6431.14,
    },
    {
      currency: 'GBP',
      rate: 4977.06,
    },
  ],
  ETH: [
    {
      currency: 'USD',
      rate: 201.76,
    },
    {
      currency: 'GBP',
      rate: 156.14,
    },
  ],
  ATL: [
    {
      currency: 'USD',
      rate: 22222,
    },
    {
      currency: 'GBP',
      rate: 9999,
    },
  ],
};

export default class AmountFormatter {
  constructor(options) {
    this.amount = options.amount;
    this.coin = options.coin;
    this.currency = options.currency;
    this.format = options.format;
    this.withCurrencySymbol = options.withCurrencySymbol;
    this.prependPlusOrMinus = options.prependPlusOrMinus;
  }

  /**
   * Formats a Number and returns a String
   * @return {String}
   */
  getFormatted() {
    let { amount } = this;
    if (this.coin && this.currency) {
      amount = this.coinToCurrency(this.amount, this.coin, this.currency);
    }

    let formatted = `${numeral(Math.abs(amount)).format(this.format)}`;
    if (this.currency) formatted = `${this.currency.symbol}${formatted}`;

    if (this.prependPlusOrMinus) {
      if (amount < 0) return `- ${formatted}`;
      if (amount > 0) return `+ ${formatted}`;
    }

    return formatted;
  }

  /**
   * Converts a coin into a currency
   * @return {Number}
   */
  coinToCurrency() {
    const { rate } = exchangeRates[this.coin].find(item => item.currency === this.currency.code);
    if (rate) return this.amount * rate;
    return this.amount;
  }
}
