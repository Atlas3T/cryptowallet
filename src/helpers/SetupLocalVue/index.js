import Vuex from 'vuex';
import { createLocalVue } from '@vue/test-utils';
import VueRouter from 'vue-router';
import VueI18n from 'vue-i18n';
import messages from '@/i18n/';
import axios from 'axios';
import { sync } from 'vuex-router-sync';
import errorHandlerPlugin from '@/boot/ErrorHandler';
import toasterPlugin from '@/boot/Toaster';
import permissionsPlugin from '@/boot/Permissions';
import appInvitationPlugin from '@/boot/AppInvitation';
import smsPlugin from '@/boot/Sms';
import emailPlugin from '@/boot/Email';
import contactsImportPlugin from '@/boot/ContactsImport';
import * as All from 'quasar'

const { Quasar, date } = All;

const components = Object.keys(All).reduce((object, key) => {
  const val = All[key]
  if (val && val.component && val.component.name != null) {
    object[key] = val
  }
  return object
}, {})


const localVue = createLocalVue();

localVue.use(Quasar, { components });
localVue.use(VueI18n);
localVue.use(Vuex);
// localVue.use(VueRouter);
localVue.prototype.$axios = axios;
errorHandlerPlugin({ Vue: localVue });
toasterPlugin({ Vue: localVue });
permissionsPlugin({ Vue: localVue });
appInvitationPlugin({ Vue: localVue });
smsPlugin({ Vue: localVue });
emailPlugin({ Vue: localVue });
contactsImportPlugin({ Vue: localVue });

/**
 * creates a new router and syncs it with the store using vuex-router-sync
 * @param store
 * @returns router
 */
const createRouter = (store = null) => {
  const router = new VueRouter({
    routes: [
      {
        path: '/contact/:id',
      },
    ],
  });

  if (store) { sync(store, router); }
  return router;
};

const i18n = new VueI18n({ messages, locale: 'en-gb' });

export {
  localVue,
  createRouter,
  i18n,
};
