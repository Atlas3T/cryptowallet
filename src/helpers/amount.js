/**
 * @TODO James Konrad
 * mocked data below
 */

import numeral from 'numeral';

const currencySymbols = {
  USD: '$',
  GPB: 'Â£',
};

const exchangeRates = {
  BTC: [
    {
      currency: 'USD',
      rate: 6431.14,
    },
    {
      currency: 'GPB',
      rate: 4977.06,
    },
  ],
  ETH: [
    {
      currency: 'USD',
      rate: 201.76,
    },
    {
      currency: 'GPB',
      rate: 156.14,
    },
  ],
  ATL: [
    {
      currency: 'USD',
      rate: 22222,
    },
    {
      currency: 'GPB',
      rate: 9999,
    },
  ],
};

export default class Amount {
  constructor(options) {
    this.amount = options.amount;
    this.coin = options.coin;
    this.currency = options.currency;
    this.format = options.format;
    this.withCurrencySymbol = options.withCurrencySymbol;
    this.prependPlusOrMinus = options.prependPlusOrMinus;
  }

  /**
   * Formats a Number and returns a String
   * @return {String}
   */
  get formatted() {
    let { amount } = this;
    if (this.coin && this.currency) {
      amount = this.coinToCurrency(this.amount, this.coin, this.currency);
    }

    let formatted = `${numeral(Math.abs(amount)).format(this.format)}`;
    if (this.currency) formatted = `${currencySymbols[this.currency]}${formatted}`;

    if (this.prependPlusOrMinus) {
      if (amount < 0) return `&#45; ${formatted}`;
      if (amount > 0) return `&#43; ${formatted}`;
    }

    return formatted;
  }

  /**
   * Converts a coin into a currency
   * @return {Number}
   */
  coinToCurrency() {
    const { rate } = exchangeRates[this.coin].find(item => item.currency === this.currency);
    return this.amount * rate;
  }
}
