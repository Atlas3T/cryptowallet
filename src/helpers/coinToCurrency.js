const exchangeRates = {
  btc: [
    {
      currency: 'USD',
      value: 6431.14,
    },
    {
      currency: 'GPB',
      value: 4977.06,
    },
  ],
  eth: [
    {
      currency: 'USD',
      value: 201.76,
    },
    {
      currency: 'GPB',
      value: 156.14,
    },
  ],
  atlas: [
    {
      currency: 'USD',
      value: 22222,
    },
    {
      currency: 'GPB',
      value: 9999,
    },
  ],
};

/**
 * Calcuates the value of a coin in a given currency
 * @param  {Number} amount
 * @param  {String} coin
 * @param  {String} currency
 * @return {Number}
 */

export default function coinToCurrency(amount, coin, currency) {
  const { value } = exchangeRates[coin].find(item => item.currency === currency);
  return amount * value;
}
