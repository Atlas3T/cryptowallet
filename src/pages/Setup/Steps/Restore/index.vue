<template>
  <div>
    <div class="container">
      <div class="restore-header">
        <h1 class="setup">{{ $t('restoreAccount') }}</h1>
        <p class="setup"> {{ $t('restoreDescription') }}</p>
      </div>

      <div class="restore-input-wrapper">
        <q-input
          v-model="seedPhrase"
          type="textarea"
          color="blueish"
          float-label="Backup Phrase"
          rows="1"
          inverted
        />
      </div>
    </div>
    <div class="btns-wrapper">
      <q-btn
        color="yellow"
        text-color="blueish"
        label="Next"
        @click="validate"
      />
    </div>
  </div>
</template>

<script>
import { mapState } from 'vuex';
import bip39 from 'bip39';

export default {
  name: 'Restore',
  data() {
    return {
      seedPhrase: '',
    };
  },
  computed: {
    ...mapState({
      id: state => parseInt(state.route.params.id, 10),
    }),
  },
  methods: {
    validate() {
      const seedPhrase = this.seedPhrase.trim().split(' ');

      if (seedPhrase.length !== 12) {
        this.$toast.create(10, this.$t('notEnoughWords'), 500);
        return false;
      }

      if (!bip39.validateMnemonic(seedPhrase.join(' '))) {
        this.$toast.create(10, 'Invalid seed phrase', 500);
        return false;
      }

      const accounts = this.$store.getters['entities/account/query']().get();
      const seedAlreadyInUse = accounts.find(account => account.seed === seedPhrase);

      if (seedAlreadyInUse) {
        this.$toast.create(10, this.$t('accountAlreadyImported'), 500);
        return false;
      }
      this.$store.dispatch('setup/setSeed', seedPhrase);
      this.$router.push({ path: `/setup/${this.id + 1}` });
      return true;
    },
  },
};

</script>

<style scoped>
.restore-input-wrapper {
  margin-top: 1rem;
  padding: 0 1em;
}
</style>
