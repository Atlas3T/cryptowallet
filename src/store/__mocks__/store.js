import Vue from 'vue';
import Vuex from 'vuex';
import VuexORM from '@vuex-orm/core';
import VuexORMLoki from '@atlascity/vuex-orm-plugin-lokijs';

import wallet from '@/store/wallet';
import contacts from '@/store/contacts/__mocks__/contacts.js';
import search from '@/store/search/__mocks__/search.js';
import settings from '@/store/settings/__mocks__/settings.js';
import setup from '@/store/setup/__mocks__/setup.js';
import qrcode from '@/store/qrcode/__mocks__/qrcode.js';
import modals from '@/store/modals/__mocks__/modals.js';

// entities
import Account from '@/store/wallet/entities/account';
import Address from '@/store/wallet/entities/address';
import Tx from '@/store/wallet/entities/tx';
import Utxo from '@/store/wallet/entities/utxo';
import Wallet from '@/store/wallet/entities/wallet';
import Coin from '@/store/wallet/entities/coin';
import KeyPair from '@/store/wallet/entities/keyPair';
import Prices from '@/store/prices';
import LatestPrice from '@/store/latestPrice';

Vue.use(Vuex);

// Setup ORM database.
const database = new VuexORM.Database();

database.register(Account, {});
database.register(Address, {});
database.register(Tx, {});
database.register(Utxo, {});
database.register(Wallet, {});
database.register(Coin, {});
database.register(KeyPair, {});
database.register(Prices, {});
database.register(LatestPrice, {});

Vue.prototype.encryptedModels = [
  Account,
  Wallet,
];

const options = {
  env: 'browser',
};

VuexORM.use(VuexORMLoki, { database, options });

export const actions = {};
export const getters = {};
export const mutations = {};
export const state = {};

export function createMocks(custom = {
  getters: {}, mutations: {}, actions: {}, state: {},
}) {
  /**
   * mock store modules.
   */
  const modules = {
    wallet,
    search,
    settings,
    contacts,
    setup,
    qrcode,
    modals,
  };

  let mockGettersCombined = {};
  let mockMutationsCombined = {};
  let mockActionsCombined = {};
  let mockStateCombined = {};

  Object.entries(modules).forEach(([name, module]) => {
    mockGettersCombined = Object.assign({}, mockGettersCombined, module.getters);
    mockMutationsCombined = Object.assign({}, mockMutationsCombined, module.mutations);
    mockActionsCombined = Object.assign({}, mockActionsCombined, module.actions);
    mockStateCombined = Object.assign({}, mockStateCombined, module.state);

    if (custom.state && custom.state[name]) {
      module.state = Object.assign(module.state, custom.state[name]);
    }

    modules[name] = {
      namespaced: true,
      getters: Object.assign({}, getters, module.getters),
      mutations: Object.assign({}, mutations, module.mutations),
      actions: Object.assign({}, actions, module.actions),
      state: Object.assign({}, state, module.state),
    };
  });

  mockGettersCombined = Object.assign({}, mockGettersCombined, custom.getters);
  mockMutationsCombined = Object.assign({}, mockMutationsCombined, custom.mutations);
  mockActionsCombined = Object.assign({}, mockActionsCombined, custom.actions);
  mockStateCombined = Object.assign({}, mockStateCombined, custom.state);
  return {
    getters: mockGettersCombined,
    mutations: mockMutationsCombined,
    actions: mockActionsCombined,
    state: mockStateCombined,
    store: new Vuex.Store({
      getters: mockGettersCombined,
      mutations: mockMutationsCombined,
      actions: mockActionsCombined,
      state: mockStateCombined,
      modules,
      plugins: [VuexORM.install(database)],
    }),
  };
}

export const { store } = createMocks();
