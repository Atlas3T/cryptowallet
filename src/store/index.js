/*eslint-disable*/

import Vue from 'vue';
import Vuex from 'vuex';
import VuexPersistence from 'vuex-persist';
import VuexORM from '@vuex-orm/core';
import VuexORMLoki from 'vuex-orm-lokijs';

// import entities.
import Account from './wallet/entities/account';
import Address from './wallet/entities/address';
import Tx from './wallet/entities/tx';
import Utxo from './wallet/entities/utxo';
import Wallet from './wallet/entities/wallet';
import Coin from './wallet/entities/coin';
import KeyPair from './wallet/entities/keyPair';
import Prices from './prices';
import LatestPrice from './latestPrice';

// import modules.
import payments from './payments';
import search from './search';
import settings from './settings';
import setup from './setup';
import contacts from './contacts';

Vue.use(Vuex);

// Setup ORM database.
const database = new VuexORM.Database();

database.register(Account, {});
database.register(Address, {});
database.register(Tx, {});
database.register(Utxo, {});
database.register(Wallet, {});
database.register(Coin, {});
database.register(KeyPair, {});
database.register(Prices, {});
database.register(LatestPrice, {});

Vue.prototype.encryptedModels = [
  Account,
];

window.Tx = Tx;
window.Utxo = Utxo;
window.Address = Address;
window.Wallet = Wallet;
window.LatestPrice = LatestPrice;
window.Prices = Prices;
window.Account = Account;
window.Coin = Coin;

// Setup persistant storage.
const vuexLocal = new VuexPersistence({
  supportCircular: true,
  storage: window.localStorage,
  modules: [settings],
});


const options = {
  env: 'browser',
};

function hydrationCompletedCallback() {
  setTimeout(() => {
    store.dispatch('settings/setLoading', false);
  }, 1000);
}

VuexORM.use(VuexORMLoki, { database, options, hydrationCompletedCallback });


/**
 * Create CryptoWallet Vuex store obj.
 * @type {Store<any>}
 */
const store = new Vuex.Store({
  modules: {
    payments,
    search,
    settings,
    setup,
    contacts,
  },
  plugins: [VuexORM.install(database)],
});

Vue.prototype.activeWallets = {};

// @todo remove line below (used for development purposes only)
window.store = store;

export default store;
