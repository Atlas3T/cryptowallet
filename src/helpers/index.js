import AmountFormatter from './AmountFormatter';
import dateTranslater from './DateTranslater';
import Contacts from './Contacts';

export {
  AmountFormatter,
  dateTranslater,
  Contacts,
};
