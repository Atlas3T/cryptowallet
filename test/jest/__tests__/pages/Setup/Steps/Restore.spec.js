import { mount } from '@vue/test-utils';
import Restore from '@/pages/Setup/Steps/Restore';
import { localVue, i18n, createRouter } from '@/helpers/SetupLocalVue';
import { createMocks as createStoreMocks } from '@/store/__mocks__/store.js';
import bip39 from 'bip39';

describe(' Restore.vue', () => {
  let wrapper;
  let store;
  let storeMocks;
  let router;
  let textInput;
  let nextButton;
  const delay = 500;


  localVue.use(bip39);

  function wrapperInit(options) {
    return mount(Restore, options);
  }

  function storeInit(custom) {
    storeMocks = createStoreMocks(custom);
    router = createRouter(storeMocks.store);
    router.push({ path: '/setup/2' });
    wrapper = wrapperInit({
      i18n,
      router,
      localVue,
      store: storeMocks.store,
      mocks: {
        bip39,
      },
    });
    store = wrapper.vm.$store;
    textInput = wrapper.find('.q-field__native');
    nextButton = wrapper.findAll('button').at(0);
  }
  beforeEach(() => { storeInit(); });

  it('renders and matches snapshot', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('validates if an entered mnemonic is the required length', () => {
    wrapper.vm.$toast.create = jest.fn();
    const invalidSeed = 'real debate another phone response toddler fee offer bundle crack monster';
    textInput.setValue(invalidSeed);
    nextButton.trigger('click');
    expect(wrapper.vm.$toast.create).toHaveBeenCalledWith(10, wrapper.vm.$t('notEnoughWords'), delay);
  });

  it('validates if an entered mnemonic is valid', () => {
    wrapper.vm.$toast.create = jest.fn();
    const invalidSeed = 'real debate another phone response toddler fee offer bundle crack monster eartw';
    textInput.setValue(invalidSeed);
    nextButton.trigger('click');
    expect(wrapper.vm.$toast.create).toHaveBeenCalledWith(10, wrapper.vm.$t('invalidSeedPhrase'), delay);
  });

  it('adds mnemonic to the store and goes to next setup step if valid', () => {
    const validSeed = 'real debate another phone response toddler fee offer bundle crack monster earth';
    textInput.setValue(validSeed);
    nextButton.trigger('click');
    expect(storeMocks.actions.setSeed).toHaveBeenCalled();
    expect(store.state.route.path).toBe('/setup/4');
  });
});
