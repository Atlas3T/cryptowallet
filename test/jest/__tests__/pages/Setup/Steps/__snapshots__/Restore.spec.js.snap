// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` Restore.vue renders and matches snapshot 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="restore-header"
    >
      <h1
        class="setup"
      >
        
        Enter Your Recovery Phrase
      
      </h1>
       
      <p
        class="setup"
      >
        
        Type in your recovery phrase (usually 12 words)
      
      </p>
    </div>
     
    <div
      class="restore-input-wrapper"
    >
      <div
        class="q-field row no-wrap items-start q-textarea q-field--outlined q-field--dark"
        float-label="Backup Phrase"
      >
        <div
          class="q-field__inner relative-position col self-stretch column justify-center"
        >
          <div
            class="q-field__control relative-position row no-wrap text-primary"
            tabindex="-1"
          >
            <div
              class="q-field__control-container col relative-position row no-wrap q-anchor--skip"
            >
              <textarea
                class="q-field__native"
                float-label="Backup Phrase"
                rows="6"
                tabindex="0"
                type="textarea"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
   
  <div
    class="btns-wrapper"
  >
    <button
      class="q-btn inline q-btn-item non-selectable q-btn--standard q-btn--rectangle bg-yellow text-blueish q-focusable q-hoverable"
      tabindex="0"
      type="button"
    >
      <div
        class="q-focus-helper"
        tabindex="-1"
      />
      <div
        class="q-btn__content text-center col items-center q-anchor--skip justify-center row"
      >
        <div>
          Next
        </div>
      </div>
    </button>
  </div>
</div>
`;
